capi-configuration/README.md

```diff
@@ -0,0 +1,333 @@
+
+## Git-Ops approach
+
+We have already taken steps to manage configuration for our components in management clusters through open source gitops tools.
+This has set a precedent for gitops management for configuration in other areas as well.
+
+Our customers generally desire to work with an `Infrastructure As Code` approach which is extendended by currently available open source gitops tooling.
+
+**This RFC will exclusively deal with the impact of using a gitops driven approach in the following sections.**
+
+## Configuration structure
+
+There are many different possible ways to structure the configuration which is used to create the clusters.
+The goal here is to provide one way that we think _can_ work and then iterating on it - we do not want to box ourselves in.
+
+### Cluster-api fundamental structure
+
+There are some links between and some fields in `cluster-api` custom resources which are essential.
+These fields can not be inferred from a webhook without context knowledge.
+It is therefor necessary to supply these values _immediately_ when the custom resource is created (e.g. `kubectl-gs` or `happa` needs to fill them).
+
+We can look at a valid `MachinePool` custom resource to determine its fundamental structure:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: capa-mp-0
+spec:
+  clusterName: capa
+  replicas: 2
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfig
+          name: capa-mp-0
+      clusterName: capa
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: AWSMachinePool
+        name: capa-mp-0
+      version: v1.16.8
+```
+- We can not infer `apiVersion` as it's an essential field.
+- We can not infer `kind` as it's an essential field.
+- We can not infer `metadata.name` as it's required to create the custom resource.
+- We can not infer `spec.clusterName` as it's unique to the created cluster.
+- We can not infer `spec.template.spec.bootstrap.configRef.name` as it's unique to the cluster.
+- We can not infer `spec.template.spec.infrastructureRef.configRef.name` as it's unique to the cluster.
+
+All other values can either be inferred from other information in the custom resource or from default values.
+
+Therefor the **minimal** custom resource which would have to be submitted by a client should be:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: capa-mp-0
+spec:
+  clusterName: capa
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          name: capa-mp-0
+      infrastructureRef:
+        name: capa-mp-0
+```
+
+This minimal custom resource can also be represented as a template which will remain valid throughout the lifecycle of `v1alpha3` in this case:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: {{clusterID}}-mp-0
```

Comment:
not sure what the template is for, but could/should the index in the end not also be templated?

Reply:
Yes - I just decided to keep it kind of simple here. Idea would be that e.g. happa or kubectl-gs uses this template

Reply:
yeah, that's why I thought the MP ID should be templated, too, so the clients could create several of them at the same time

Reply:
Currently `kubectl-gs` templates the machinepool name simply as the nodepool id in the same way as we do it for our current clusters.

Reply:
Makes sense - I will change this as I took it from upstream templates.

Reply:
Done.